src/main/java/com/jcabi/aspects/aj/MethodCacher.java

```diff
@@ -327,21 +334,25 @@ public synchronized Object through() throws Throwable {
                         Mnemos.toText(
                             method, this.point.getArgs(), true, false
                         ),
-                        Mnemos.toText(this.cached, true, false),
+                        Mnemos.toText(this.cached.get(), true, false),
                         System.currentTimeMillis() - start,
                         suffix
                     );
                 }
                 this.executed = true;
             }
-            return this.key.through(this.cached);
+            return this.key.through(this.cached.get());
         }
         /**
          * Is it expired already?
          * @return TRUE if expired
          */
         public boolean expired() {
-            return this.executed && this.lifetime < System.currentTimeMillis();
+            final boolean expired = this.lifetime < System.currentTimeMillis();
+            final boolean collected = this.executed
+                    && this.hasresult
+                    && this.cached.get() == null;
+            return this.executed && (expired || collected);
```

Comment:
@nesterov-n looks like pretty complex logic right now. I'm afraid we would need a test, simplification or some other proof that it works correctly